src/routes/mod_creation.rs

```diff
@@ -0,0 +1,401 @@
+            CreateError::SerDeError(..) => StatusCode::BAD_REQUEST,
+            CreateError::MultipartError(..) => StatusCode::BAD_REQUEST,
+            CreateError::InvalidUtf8Input(..) => StatusCode::BAD_REQUEST,
+            CreateError::MissingValueError(..) => StatusCode::BAD_REQUEST,
+        }
+    }
+
+    fn error_response(&self) -> HttpResponse {
+        HttpResponse::build(self.status_code()).json(ApiError {
+            error: match self {
+                CreateError::EnvError(..) => "environment_error",
+                CreateError::DatabaseError(..) => "database_error",
+                CreateError::FileHostingError(..) => "file_hosting_error",
+                CreateError::SerDeError(..) => "invalid_input",
+                CreateError::MultipartError(..) => "invalid_input",
+                CreateError::InvalidUtf8Input(..) => "invalid_input",
+                CreateError::MissingValueError(..) => "invalid_input",
+            },
+            description: &self.to_string(),
+        })
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+struct InitialVersionData {
+    pub file_indexes: Vec<i32>,
+    pub version_slug: String,
+    pub version_title: String,
+    pub version_body: String,
+    pub dependencies: Vec<ModId>,
+    pub game_versions: Vec<GameVersion>,
+    pub version_type: VersionType,
+    pub loaders: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+struct ModCreateData {
+    /// The title or name of the mod.
+    pub mod_name: String,
+    /// The namespace of the mod
+    pub mod_namespace: String,
+    /// A short description of the mod.
+    pub mod_description: String,
+    /// A long description of the mod, in markdown.
+    pub mod_body: String,
+    /// A list of initial versions to upload with the created mod
+    pub initial_versions: Vec<InitialVersionData>,
+    /// The team of people that has ownership of this mod.
+    pub team_members: Vec<TeamMember>,
+    /// A list of the categories that the mod is in.
+    pub categories: Vec<String>,
+    /// An optional link to where to submit bugs or issues with the mod.
+    pub issues_url: Option<String>,
+    /// An optional link to the source code for the mod.
+    pub source_url: Option<String>,
+    /// An optional link to the mod's wiki page or other relevant information.
+    pub wiki_url: Option<String>,
+}
+
+#[post("api/v1/mod")]
+pub async fn mod_create(
+    mut payload: Multipart,
+    client: Data<Client>,
+    upload_url: Data<UploadUrlData>,
+) -> Result<HttpResponse, CreateError> {
+    //TODO Switch to transactions for safer database and file upload calls (once it is implemented in the APIs)
+    let cdn_url = dotenv::var("CDN_URL")?;
+
+    let db = client.database("modrinth");
+
+    let mods = db.collection("mods");
+    let versions = db.collection("versions");
+
+    let mut mod_id = ModId(random_base62(8));
+
+    //Check if ID is unique
+    loop {
```

Comment:
We may want to add a second condition that it can't take more than 10 or 20 cycles to prevent a potential infinite loop.  If it fails, log an error and break out so that the server doesn't stall.

Reply:
Good point - I'll do this real quick as a safety measure.